docs/ql-style-guide.md

```diff
@@ -54,6 +54,7 @@ select c, "This call to '$@' is deprecated because " + reason + ".",
    - Immediately after `if`, `then`, or `else` keywords. The `then` and `else` parts *should* be consistent.
 1. *Avoid* other line breaks in declarations, other than to break long lines.
 1. When operands of *binary operators* span two lines, the operator *should* be placed at the end of the first line.
+1. If the parameter list needs to be broken across multiple lines then there must *must* be a line break after the opening `(`, the parameter declarations indented one level, and the `) {` on its own line at the outer indentation.
```

Comment:
The other option that was discussed internally was to double-indent parameter lists:

```
predicate methodStats(
    string qualifiedName, string name, int numberOfParameters,
    int numberOfStatements, int numberOfExpressions, int linesOfCode,
    int nestingDepth, int numberOfBranches) {
  ...
} 
```

To decide between the two options, I'd like to see how you propose to handle long `extends` lists and ideally also long `instanceof` and `implements` lists. Double-indenting also works for those:

```
abstract class TranslatedListInitialization extends
    TranslatedInitialization, InitializationContext {
  TranslatedListInitialization() { ... }
}
```

I'm not sure whether `extends` should be last on a line. I think the Elephants are serious about the `instanceof`/`implements` proposal, so a solution should ideally work for that as well. Here's an attempt to format an example from that proposal (adapted to use `or` instead of `;` because it seems like we'll go with that):

```
class CallExpr
    extends Expr
    instanceof LambdaCall or MethodCall {
  ...
}
```

I don't think that looks very good. I'd rather have a single indent and place `{` alone on a line. But can we then justify that it should be `) {` alone and not `{` alone for parameter lists?

Reply:
I agree with @jbj here. The long line splitting problem isn't just about parameter lists. We should pick a solution that works for both brace/paren-enclosed lists like parameter lists, and for syntax that looks like a base class list where there are no enclosing braces. A straightforward rule of "continuation lines must be double indented" works for both cases. So would "An opening brace always goes on its own line".

Reply:
I think that "opening brace on its own line after a long extends list" is consistent with "`) {` on its own line after a long parameter list.  In both cases the opening brace is moved to the following line and if the in-between stuff that's long is wrapped in parentheses then those parentheses are placed in the way that we usually place parentheses for everything else (e.g. `exists`).  In any case, I don't have any particular opinion about the best formatting for long extends-lists.

Reply:
Long argument lists are probably more common than long parameter lists and extends lists, and I think those should also follow the same rules. How about this?

```java
myPredicate(
  longArgument1, longArgument2,
  longArgument3, longArgument4
) =
  myOtherPredicate(
    longArgument1, longArgument2,
    longArgument3, longArgument4
  )
```

This looks good to me and is consistent with @aschackmull's proposed formatting of parameter lists, I think.